src/groupeddataframe/splitapplycombine.jl

```diff
@@ -169,48 +169,59 @@ function groupby(df::AbstractDataFrame, cols;
-  for each group as the length of the returned vector for that group.
-- A data frame, a named tuple of vectors or a matrix gives a `DataFrame` with the same
-  additional columns and as many rows for each group as the rows returned for that group.
-  As a special case, returning an empty table with zero columns is allowed,
-  whatever the number of columns returned for other groups.
-
-`f` must always return the same kind of object (a single value or a table as defined
-in the above list) for all groups, and with the same column names (defined explicitly
-or auto-generated). If an empty vector, data frame, or named tuple is returned then
-the corresponding group is dropped from the result.
-In particular, named tuples cannot mix single values and vectors.
-Due to type instability, returning a single value or a named tuple is dramatically
-faster than returning a data frame.
-
-Optimized methods are used when standard summary functions (`sum`, `prod`,
-`minimum`, `maximum`, `mean`, `var`, `std`, `first`, `last` and `length`)
-are specified using the `Pair` syntax (e.g. `col => sum`).
-When computing the `sum` or `mean` over floating point columns, results will be less
-accurate than the standard [`sum`](@ref) function (which uses pairwise summation). Use
-`col => x -> sum(x)` to avoid the optimized method and use the slower, more accurate one.
-
-Column names are automatically generated when necessary using the names `x1`, `x2` and so on
-if `f` is a `Base.Callable` or an `AbstractMatrix` is returned by it
-and using the rules defined in [`select`](@ref) otherwise.
-"""
+    """
+    `fun` or `f` in `pair` can return a single value, a row or multiple rows.
+    The type of the returned value determines the shape of the resulting `DataFrame`.
+    There are four kind of return values allowed:
+    - A single value gives a `DataFrame` with a single additional column and one row per group.
+    - A named tuple of single values or a [`DataFrameRow`](@ref) gives a `DataFrame` with
+      one additional column for each field and one row per group (returning a named tuple
+      will be faster).
+    - A vector gives a `DataFrame` with a single additional column and as many rows
+      for each group as the length of the returned vector for that group.
+    - A data frame, a named tuple of vectors or a matrix gives a `DataFrame` with
+      the same additional columns and as many rows for each group as the rows
+      returned for that group (returning a named tuple is the fastest option).
+      Returning a table with zero columns is allowed, whatever the number of columns
+      returned for other groups.
+
+    It is not allowed to mix single values and vectors if a named tuple is returned.
+
+    `fun` or `f` in `pair` must always return the same kind of object (out of four
+    kinds defined above) for all groups, and with the same column names if return
+    value specifies them.
```

Comment:
Maybe this goes without saying? Anyway, if names are generated automatically, they also have to be the same because the number of columns must be equal.
```suggestion
    kinds defined above) for all groups, and with the same column names.
```